File: src/main/Leetcode00019.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Leetcode00019 {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode node = head;
        int length = 0;
        while (node != null) {
            node = node.next;
            length++;
        }
        int cut = length - n - 1;
        if (cut < 0) {
            return head.next;
        }
        node = head;
        while (node.next != null && cut-- > 0) {
            node = node.next;
        }
        node.next = node.next.next;
        return head;
    }
}

File: src/main/Leetcode00019.swift
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.next = nil
 *     }
 * }
 */
class Solution {
    func count(head: ListNode?) -> Int {
        return head != nil ? (1 + count(head!.next)) : 0
    }
    func removeNthFromEnd(head: ListNode?, _ n: Int) -> ListNode? {
        let c = count(head)
        var l = c - n - 1
        if l == -1 {
            let newhead = head?.next
            head?.next = nil
            return newhead
        } else {
            var node = head
            for _ in 0.stride(to:l,by:1) {
                node = node!.next
            }
            node!.next = node!.next!.next
        }
        return head
    }
}
